spiel/cmd.c

```diff
@@ -2000,6 +2047,269 @@ static int spiel__proc_is_to_update_stats(struct spiel_proc_item *proc,
+	}
+
+	memcpy(proc->sci_data.pd_key.b_addr, rep->sspr_bckey.b_addr,
+		rep->sspr_bckey.b_nob);
+	memcpy(proc->sci_data.pd_rec.b_addr, rep->sspr_bcrec.b_addr,
+		rep->sspr_bcrec.b_nob);
+	proc->sci_data.pd_kv_count = rep->sspr_kv_count;
+
+leave:
+	m0_rpc_machine_lock(proc->sci_spc->spc_rmachine);
+	m0_fop_put(&proc->sci_fop);
+	m0_fop_fini(&proc->sci_fop);
+	m0_rpc_machine_unlock(proc->sci_spc->spc_rmachine);
+
+	conn_timeout = m0_time_from_now(SPIEL_CONN_TIMEOUT, 0);
+	m0_rpc_link_disconnect_sync(&proc->sci_rlink, conn_timeout);
+	m0_rpc_link_fini(&proc->sci_rlink);
+	proc->sci_rc = rc;
+	m0_semaphore_up(&proc->sci_barrier);
+	M0_LEAVE();
+}
+
+static struct m0_sm_group* spiel_counter_sm_group(struct spiel_proc_counter_item *i)
+{
+	return i->sci_spc->spc_confc->cc_group;
+}
+
+static void spiel_process_counter_replied(struct m0_rpc_item *item)
+{
+	struct m0_fop                  *fop  = m0_rpc_item_to_fop(item);
+	struct spiel_proc_counter_item *proc = M0_AMB(proc, fop, sci_fop);
+
+	M0_ENTRY(FID_F, FID_P(&proc->sci_fid));
+	proc->sci_ast.sa_cb = spiel_process_counter_replied_ast;
+	m0_sm_ast_post(spiel_counter_sm_group(proc), &proc->sci_ast);
+	M0_LEAVE();
+}
+
+struct m0_rpc_item_ops spiel_process_counter_ops = {
+	.rio_replied = spiel_process_counter_replied
+};
+
+static bool spiel_proc_counter_item_rlink_cb(struct m0_clink *clink)
+{
+	struct spiel_proc_counter_item *proc = M0_AMB(proc, clink, sci_rlink_wait);
+	struct m0_ss_process_req       *req;
+	struct m0_fop                  *fop = &proc->sci_fop;
+	struct m0_rpc_item             *item;
+	m0_time_t                       conn_timeout;
+	int                             rc;
+
+	M0_ENTRY(FID_F, FID_P(&proc->sci_fid));
+	m0_clink_fini(clink);
+	if (proc->sci_rlink.rlk_rc != 0) {
+		M0_LOG(M0_ERROR, "connect failed");
+		m0_semaphore_up(&proc->sci_barrier);
+		M0_LEAVE();
+		return true;
+	}
+	m0_fop_init(fop, &m0_fop_process_fopt, NULL, m0_fop_release);
+	rc = m0_fop_data_alloc(fop);
+	if (rc != 0) {
+		M0_LOG(M0_ERROR, "fop data alloc failed");
+		goto fop_fini;
+	}
+
+	fop->f_item.ri_rmachine = proc->sci_spc->spc_rmachine;
+	req                     = m0_ss_fop_process_req(fop);
+	req->ssp_cmd            = M0_PROCESS_COUNTER;
+	req->ssp_id             = proc->sci_fid;
+	item                    = &fop->f_item;
+	item->ri_ops            = &spiel_process_counter_ops;
+	item->ri_session        = &proc->sci_rlink.rlk_sess;
+	item->ri_prio           = M0_RPC_ITEM_PRIO_MID;
+	item->ri_nr_sent_max    = 5;
+	m0_fop_get(fop);
+	rc = m0_rpc_post(item);	
+	if (rc != 0) {
+		M0_LOG(M0_ERROR, "rpc post failed");
+		goto fop_put;
+	}
+
+	proc->sci_rc = rc;
+	
+	M0_LEAVE();
+	return true;
+fop_put:
+	m0_fop_put_lock(fop);
+fop_fini:
+	m0_fop_fini(fop);
+	conn_timeout = m0_time_from_now(SPIEL_CONN_TIMEOUT, 0);
+	m0_rpc_link_disconnect_sync(&proc->sci_rlink, conn_timeout);
+	m0_rpc_link_fini(&proc->sci_rlink);
+	proc->sci_rc = rc;
+	m0_semaphore_up(&proc->sci_barrier);
+	M0_LEAVE();
+	return true;
+}
```

Comment:
[Suggestion] In the above function, return and goto statements are interleaved at multiple places.
Do you think it can be simplified? something like:
```
	if (proc->sci_rlink.rlk_rc != 0) {
		M0_LOG(M0_ERROR, "connect failed");
		m0_semaphore_up(&proc->sci_barrier);
		goto ret;
	}
	m0_fop_init(fop, &m0_fop_process_fopt, NULL, m0_fop_release);
	rc = m0_fop_data_alloc(fop);
	if (rc != 0) {
		M0_LOG(M0_ERROR, "fop data alloc failed");
		goto fop_fini;
	}

	fop->f_item.ri_rmachine = proc->sci_spc->spc_rmachine;
	req                     = m0_ss_fop_process_req(fop);
	req->ssp_cmd            = M0_PROCESS_COUNTER;
	req->ssp_id             = proc->sci_fid;
	item                    = &fop->f_item;
	item->ri_ops            = &spiel_process_counter_ops;
	item->ri_session        = &proc->sci_rlink.rlk_sess;
	item->ri_prio           = M0_RPC_ITEM_PRIO_MID;
	item->ri_nr_sent_max    = 5;
	m0_fop_get(fop);
	rc = m0_rpc_post(item);
	if (rc != 0) {
		M0_LOG(M0_ERROR, "rpc post failed");
		m0_fop_put_lock(fop);
	}

	proc->sci_rc = rc;
	goto ret;

fop_fini:
	m0_fop_fini(fop);
	conn_timeout = m0_time_from_now(SPIEL_CONN_TIMEOUT, 0);
	m0_rpc_link_disconnect_sync(&proc->sci_rlink, conn_timeout);
	m0_rpc_link_fini(&proc->sci_rlink);
	proc->sci_rc = rc;
	m0_semaphore_up(&proc->sci_barrier);

ret:
	M0_LEAVE();
	return true;

```

Reply:
Yes, simplified the code as per your suggestion.